Fifth day

zeitschlag opened the pull request.

zeitschlag committed:
[WIP] first draft for fifth day

zeitschlag committed:
Add solution for 5.1

zeitschlag committed:
extract code into own functions

zeitschlag committed:
rename functions

zeitschlag committed:
introduce new opcodes

zeitschlag committed:
[WIP] current state for fifth day

zeitschlag committed:
Solve second puzzle of the fifth day

I had some issues with strings and integers

klyonrad commented on 2019/05/5.rb:
```diff
@@ -0,0 +1,169 @@
+	first_input_value = Integer(intcode[first_input_index])
+  end
+  
+  return first_input_value
```
```suggestion
  first_input_value
```

You do not need to (aka *should not*) write `return` at the end of ruby methods. The last thing that was evaluated **is** the return value.

Even crazier: You don't even need that variable ([Lint/UselessAssignment](https://docs.rubocop.org/en/stable/cops_lint/#lintuselessassignment))

```ruby
  if first_param_index_mode == IMMEDIATEMODE
    first_input_index
  else
    Integer(intcode[first_input_index])
  end
end
# That is all
```

I suggest that you install `rubocop` right away and at least autocorrect your code for the Lint, Layout and Style cops (`rubocop --auto-correct only Layout; rubocop -a only Style; rubocop -a only Lint`). I think it's better to do that right away while learning ruby instead of being annoyed by it later on

klyonrad replied:
Another good resource is also https://rubystyle.guide

klyonrad commented on 2019/05/5.rb:
```diff
@@ -0,0 +1,169 @@
+  end
+end
+
+def get_opcode_for_command(command)
```
That `opcode` concept looks like it has a lot of potential for becoming proper class, with subclasses and such and this method would become the *factory method*. But I guess, time's too short for "proper OO" 😉

Basically everything is controlled by these symbols, so I could imagine classes like

AddCommand, MultiplyCommand, JumpIfFalse, etc.

zeitschlag replied:
Yeah, you're probably right. I've fiddled around with classes in Ruby as you can see [here](https://github.com/zeitschlag/adventofcode/pull/3) or [here](https://github.com/zeitschlag/adventofcode/blob/master/2019/03/third_day.rb). I think, in the long run, it will totally make sense to rewrite this with classes.

Thank you for your suggestion ☺️

klyonrad commented on 2019/05/5.rb:
```diff
@@ -0,0 +1,169 @@
+  result = run_intcode(intcode)
+end
+
+first_puzzle
```
I know, there are conflicting philosophies about ordering of methods, but I would suggest you to try something:

```ruby
first_puzzle

def first_puzzle 
  intcode = IO.read("input.txt").split(",")
  result = run_intcode(intcode)
end

# and then definitions of all the other methods in the order of when they are needed

```

It is just so refreshing for the reading experience when you compare it to the "stupid" limit of compiled languages that you cannot call what is only later compiled

zeitschlag replied:
I tried that before, but obviously, I did something wrong. So I gave it a second try, but I got this error message:

```
Traceback (most recent call last):
5.rb:4:in `<main>': undefined local variable or method `first_puzzle' for main:Object (NameError)
```

Can you point me into the right direction, please?

klyonrad replied:
Okay, wow sorry for the wrong suggestion 😂

I guess the joke is that every day ruby development is rarely written outside of a class, so when you call methods in the global scope they need to be defined before 🤯

klyonrad commented on 2019/05/5.rb:
```diff
@@ -0,0 +1,169 @@
+        input[output_index] = "0"
+      end
+    else
+      puts "Weird opcode (#{opcode} at #{i}), command: #{command}, aborting..."
```
You're not really aborting 😉 

```suggestion
      raise "Weird opcode (#{opcode} at #{i}), command: #{command}, aborting..."
```

klyonrad commented on 2019/05/5.rb:
```diff
@@ -0,0 +1,169 @@
+end
+
+def get_third_parameter_value(command, intcode, index)
+  output_index = Integer(intcode[index+3])
```
Not a problem here, but just a knowledge drop: Be a bit wary with parsing integers like this. Ruby might derive the number base on what it thinks is right

```ruby
Integer('0xAF')
 => 175
```

I mention that because of the puzzle domain of op codes

klyonrad commented on 2019/05/5.rb:
```diff
@@ -0,0 +1,169 @@
+def get_opcode_for_command(command)
+  if /1$/ =~ command
+    return :add
+  elsif /2$/ =~ command
```
Is the regexp matching necessary? Because from testing around it looks here you check if the string ends with a `2`.

That could be also done with [`command.end_with?('2')`](https://ruby-doc.org/core-2.6/String.html#method-i-end_with-3F)

If it's testing more than nevermind :)

zeitschlag replied:
Obviously it's not :D I had the idea, and then I gave it a try and then I stopped looking for other solutions, although I prefer your suggestions, to be honest. It's much more readable

klyonrad replied:
Yeah that’s why I suggested it - for the readability.

If you want the ultra amount of readability, you could install and require the activesupport gem, this would give even more convenience methods for strings and Object

klyonrad commented on 2019/05/5.rb:
```diff
@@ -0,0 +1,169 @@
+  elsif opcode == :read || opcode == :write
+    return 1
+  end
+end
```
I know I said the thing about unnecessar returns already, but here it is also not necessary

```ruby
def number_of_parameters_for_opcode(opcode)
  if opcode == :halt
    0
  elsif opcode == :add || opcode == :multiply || opcode == :less_than || opcode == :equals
    3
  elsif opcode == :jump_if_false || opcode == :jump_if_true
    2
  elsif opcode == :read || opcode == :write
    1
  end
end
```

klyonrad reviewed (commented):
Just a couple of ideas for more idomatic ruby

zeitschlag commented:
@klyonrad Thank you so much for your feedback!

zeitschlag committed:
Remove returns

Ruby returns the last statement automatically. I think I have to get used to that, although I’ve seen something similar from blocks in swift

zeitschlag committed:
Abort program when invalid opcode appears

zeitschlag committed:
Replace regex with string-method
